lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+            response_object = {
+                    "status": "success", # will not present if failed
+                    "message": "Successfully logged in."/"failure message",
+                    "auth_token": "<authorization_token>", # Will not present if failed
+                }
+        """
+        # get the post data
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = post_data.get("username")
+        if not username:
+            self.logger.warning("Username not provided during the login process")
+            abort(400, message="Please provide a valid username")
+
+        password = post_data.get("password")
+        if not password:
+            self.logger.warning("Password not provided during the login process")
+            abort(400, message="Please provide a valid password")
+
+        try:
+            # fetch the user data
+            user = User.query(username=username)
+        except Exception:
+            self.logger.exception("Exception occurred during user login")
+            abort(500, message="INTERNAL ERROR")
+
+        if not user:
+            self.logger.warning(
+                "No user found in the db for Username: {} while login", username
+            )
+            abort(403, message="No such user, please register first")
+
+        # Validate the password
+        if not check_password_hash(user.password, password):
+            self.logger.warning("Wrong password for user: {} while login", username)
+            abort(401, message="Bad login")
+
+        try:
+            auth_token = self.auth.encode_auth_token(
+                self.token_expire_duration, user.id
+            )
+        except (
+            jwt.InvalidIssuer,
+            jwt.InvalidIssuedAtError,
+            jwt.InvalidAlgorithmError,
+            jwt.PyJWTError,
+        ):
+            self.logger.error(
+                "Could not encode the JWT auth token for user: {} while login", username
+            )
+            abort(
+                500, message="INTERNAL ERROR",
+            )
+
+        # check if the user is already logged in, in case the request has a an authorization header included
+        # We would still proceed to login and return a new auth token to the user
+        if user == self.auth.token_auth.current_user():
+            self.logger.warning("User already logged in and trying to re-login")
+            if self.auth.token_auth.get_auth()["token"] == auth_token:
+                # If a user attempts to re-login immediately within a millisecond,
+                # the new auth token will be same as the one present in the header
+                # Therefore we will not fulfill the re-login request
+                self.logger.warning("Attempted immediate re-login")
+                abort(403, message="Retry login after some time")
```

Comment:
My opinion was always that `login` should ignore an authorization token in the header entirely; login simply doesn't require a user to be pre-authorized, so we shouldn't care.

In theory, if we did take two login attempts in close succession and generated identical tokens, we might have a management problem down the road since the `auth_token` table would have two rows with the same token. Logging out the same token twice to get rid of them is a bit weird, though not necessarily a problem; but if we don't persist the duplicate, we have "two sessions" that will be invalidated when either logs out.

That's unlikely to be an issue in production, but multiple logins (and even in relatively quick succession) are pretty easy to do during testing so it's something we need to consider deliberately.

Without a more sensible model, I kinda like the idea that the duplicate session will fail login. And this seems to me more of a 403 (it's forbidden to create a duplicate auth token) than 409. (It's not a "conflict" in the current state of a resource.)

Reply:
It is an extremely unlikely to happen in reality that the tokens are similar

Reply:
I suggested 409/Conflict because the suggested resolution for the user is to "retry...after some time" which sounded like a resource state issue to me.

I'm good (I think...) with ignoring the auth token for a login attempt.

I don't think I have a problem with giving out identical tokens to multiple concurrent logins:  as Nikhil says, this is extremely unlikely to happen, and I don't think the downsides are that bad.  (In the case of a human user, s/he just has to log in again if s/he gets unexpectedly logged out.)  We shouldn't run into the case of duplicate database rows, because we've configured the table for uniqueness (I think) -- instead, we would get blown out with an error on the second add...which might make things interesting (we would have to catch it separately and proceed assuming that the token is valid).  I certainly wouldn't recommend reference-counting or anything which would require multiple logouts to get rid of them.  (As I said, I'm not worried about losing the second session when the first logs out.)

Reply:
At this point, this should be captured in an issue and we should move along.

FWIW, I now think that this whole problem only arises if the second login attempt (coming "nearly instantaneously" with the first) happens to include the token returned to the first login attempt.  That strikes me as being _so_ unlikely that there is no point in guarding against this.  (Specifically, where is the harm in giving the same token to the requestor who already has it?  In some contexts, that would be totally normal behavior....)

Reply:
Filed https://github.com/distributed-system-analysis/pbench/issues/2129